cmd/fleet/handleAck.go

```diff
@@ -295,3 +272,41 @@ func _getAPIKeyIDs(agent *model.Agent) []string {
 	}
 	return keys
 }
+
+// Generate an update script that validates that the policy_id
+// has not changed underneath us by an upstream process (Kibana or otherwise).
+// We have a race condition where a user could have assigned a new policy to
+// an agent while we were busy updating the old one.  A blind update to the
+// agent record without a check could set the revision and coordIdx for the wrong
+// policy.  This script should be coupled with a "retry_on_conflict" parameter
+// to allow for *other* changes to the agent record while we running the script.
+// (For example, say the background bulk check-in timestamp update task fires)
+//
+// WARNING: This assumes the input data is sanitized.
+
+const kUpdatePolicyPrefix = `{"script":{"lang":"painless","source":"if (ctx._source.policy_id == params.id) {ctx._source.` +
+	dl.FieldPolicyRevisionIdx +
+	` = params.rev;ctx._source.` +
+	dl.FieldPolicyCoordinatorIdx +
+	`= params.coord;ctx._source.` +
+	dl.FieldUpdatedAt +
+	` = params.ts;} else {ctx.op = \"noop\";}","params": {"id":"`
+
+func makeUpdatePolicyBody(policyId string, newRev, coordIdx int64) []byte {
+
+	var buf bytes.Buffer
```

Comment:
would the strings.Builder be faster?

Reply:
It is faster, but you get an extra copy because there's no Bytes() accessor.